src/masternodeman.cpp

```diff
@@ -644,6 +671,406 @@ void CMasternodeMan::ProcessMessage(CNode* pfrom, std::string& strCommand, CData
+                            pair.second.IsPoSeBanned() ? "banned" : "",
+                            pair.second.vin.prevout.ToStringShort(), pair.second.addr.ToString());
+                continue;
+            }
+            LogPrint("masternode", "CMasternodeMan::DoFullVerificationStep -- Verifying masternode %s rank %d/%d address %s\n",
+                        pair.second.vin.prevout.ToStringShort(), pair.first, nRanksTotal, pair.second.addr.ToString());
+            if(SendVerifyRequest((CAddress)pair.second.addr, vSortedByAddr) && ++nCount >= nCountMax) break;
+        } else if(pair.first > MAX_POSE_RANK) {
+            LogPrint("masternode", "CMasternodeMan::DoFullVerificationStep -- Rank %d/%d, must be in top %d to send verify request\n",
+                        pair.first, nRanksTotal, MAX_POSE_RANK);
+            break;
+        }
+    }
+    LogPrint("masternode", "CMasternodeMan::DoFullVerificationStep -- Sent verification requests to %d masternodes\n", nCount);
+}
+
+void CMasternodeMan::CheckSameAddr()
+{
+    if(!masternodeSync.IsSynced() || vMasternodes.empty()) return;
+
+    std::vector<CMasternode*> vBan;
+    std::vector<CMasternode*> vSortedByAddr;
+    std::set<CAddress> setAddrsToVerify;
+
+    size_t nMaxAddrToVerify = 10;
+
+    {
+        LOCK(cs);
+
+        CMasternode* prevMasternode;
+
+        BOOST_FOREACH(CMasternode& mn, vMasternodes) {
+            vSortedByAddr.push_back(&mn);
+        }
+
+        sort(vSortedByAddr.begin(), vSortedByAddr.end(), CompareByAddr());
+
+        BOOST_FOREACH(CMasternode* pmn, vSortedByAddr) {
+            // check only (pre)enabled masternodes
+            if(!pmn->IsEnabled() && !pmn->IsPreEnabled()) continue;
+            // initial step
+            if(!prevMasternode) {
+                prevMasternode = pmn;
+                continue;
+            }
+            // second+ step
+            if(pmn->addr == prevMasternode->addr) {
+                if(prevMasternode->IsPoSeVerified()) {
+                    vBan.push_back(pmn);
+                } else if(pmn->IsPoSeVerified()) {
+                    vBan.push_back(prevMasternode);
+                    prevMasternode = pmn;
+                } else {
+                    // neither one of them is verified, let's try now
+                    setAddrsToVerify.insert((CAddress)pmn->addr);
+                    if(setAddrsToVerify.size() >= nMaxAddrToVerify) break;
+                }
+            }
+        }
+    }
+
+    // ban duplicates
+    BOOST_FOREACH(CMasternode* pmn, vBan) {
+        pmn->nPoSeBanScore++;
+    }
+
+    BOOST_FOREACH(const CAddress &addr, setAddrsToVerify) {
+        LogPrintf("CMasternodeMan::CheckSameAddr -- Multiple Masternodes with the same IP found addr=%s\n", addr.ToString());
+        SendVerifyRequest(addr, vSortedByAddr);
+    }
+}
+
+bool CMasternodeMan::SendVerifyRequest(const CAddress& addr, const std::vector<CMasternode*>& vSortedByAddr)
+{
+    CNode* pnode = ConnectNode(addr, NULL, true);
```

Comment:
Where are these connections closed ?  If they aren't closed it seems like you will eventually end up with 4000 open outgoing connections.